src/transformers/generation/utils.py

```diff
@@ -3404,184 +3439,241 @@ def _beam_search(
+
                 if output_hidden_states:
                     decoder_hidden_states += (
-                        (outputs.decoder_hidden_states,)
+                        (model_outputs.decoder_hidden_states,)
                         if self.config.is_encoder_decoder
-                        else (outputs.hidden_states,)
+                        else (model_outputs.hidden_states,)
                     )
 
-            # reshape for beam search
-            vocab_size = next_token_scores.shape[-1]
-            next_token_scores = next_token_scores.view(batch_size, num_beams * vocab_size)
-
-            # Beam token selection: pick 1 + eos_token_id.shape[0] next tokens for each beam so we have at least 1
-            # non eos token per beam.
-            n_eos_tokens = eos_token_id.shape[0] if eos_token_id is not None else 0
-            n_tokens_to_keep = max(2, 1 + n_eos_tokens) * num_beams
+            # This is needed to properly delete logits which may be very large for first iteration
+            # Otherwise a reference to outputs is kept which keeps the logits alive in the next iteration
+            del model_outputs
+
+            log_probs = unflatten_beam_dim(log_probs, batch_size, num_beams)
+            log_probs = log_probs + running_beam_scores[:, :, None]  # (batch_size, num_beams, vocab_size)
+            vocab_size = log_probs.shape[2]
+            log_probs = torch.reshape(log_probs, (batch_size, num_beams * vocab_size))
+
+            # c. Retrieve top-K continuations -- each item in batch has `num_beams` * `vocab_size` candidate
+            # continuations. For each item, get the top K [K = (number of EOS tokens + 1) * `num_beams`] candidates
+            # with the highest log-probabilities, or sample them without replacement. We gather the top K (as opposed
+            # to `num_beams`, or any number lower than K) here so that we have at least `num_beams` sequences
+            # remaining to continue the live beam search.
+            # Note that other stopping criteria might result in impossible to continue beams, i.e. all continuations
+            # selected in this step hit the stopping criteria.
+
+            # Gather the top K scores from _all_ beams.
+            beams_to_keep = max(2, 1 + n_eos_tokens) * num_beams
             if do_sample:
-                probs = nn.functional.softmax(next_token_scores, dim=-1)
-                next_tokens = torch.multinomial(probs, num_samples=n_tokens_to_keep)
-                next_token_scores = torch.gather(next_token_scores, -1, next_tokens)
-                next_token_scores, _indices = torch.sort(next_token_scores, descending=True, dim=1)
-                next_tokens = torch.gather(next_tokens, -1, _indices)
+                topk_indices = torch.multinomial(nn.functional.softmax(log_probs, dim=-1), num_samples=beams_to_keep)
+                topk_log_probs = torch.gather(input=log_probs, dim=1, index=topk_indices)
```

Comment:
this bit is also compile compatible AFAIK!
Would be nice to split the post processing part in another fun IMO!

Reply:
on torch 2.3 `torch.multinomial` was not compilable 😞 Not sure what's the state atm

In practice, the whole beam search loop should be compilable, at least in the greedy case 🤞 there are no data-dependent conditions or slices